File: 9_排序和搜索/二分/二分答案法/二分加滑窗/410. 分割数组的最大值.ts
/**
 * @param {number[]} nums 1 <= nums.length <= 1000
 * @param {number} m
 * @return {number}
 * 给定一个非负整数数组 nums 和一个整数 m ，你需要将这个数组分成 m 个非空的连续子数组。
 * 设计一个算法使得这 m 个子数组各自和的最大值最小
 * 返回这个和
 * @summary
 * 非负整数数组:和的单调性=>二分查找
 */
const splitArray = function (nums: number[], m: number): number {
  let l = Math.max.apply(null, nums)
  let r = nums.reduce((pre, cur) => pre + cur, 0)
  while (l <= r) {
    const mid = (l + r) >> 1
    const count = countNGT(nums, mid)
    if (count > m) {
      l = mid + 1
    } else {
      r = mid - 1
    }
  }

  return l

  /**
   *
   * @param nums
   * @param mid
   * @returns 每个子数组和不超过mid 至少需要分成多少个子数组
   */
  function countNGT(nums: number[], mid: number): number {
    let res = 1
    let curSum = 0

    for (const num of nums) {
      if (num + curSum > mid) {
        curSum = num
        res++
      } else {
        curSum += num
      }
    }

    return res
  }
}

console.log(splitArray([7, 2, 5, 10, 8], 2))

export default 1

File: 9_排序和搜索/二分/二分答案法/410. 分割数组的最大值.py
# 410. 分割数组的最大值
# 给定一个非负整数数组 nums 和一个整数 k ，你需要将这个数组分成 k 个非空的连续子数组。
# !设计一个算法使得这 k 个子数组各自和的最大值最小。
# n<=1000 0<=nums[i]<=1e6


from typing import List


class Solution:
    def splitArray(self, nums: List[int], k: int) -> int:
        def check(mid: int) -> bool:
            """每段不超过mid,是否能分成<=k段"""
            count = 1
            curSum = 0
            for num in nums:
                if num > mid:
                    return False
                curSum += num
                if curSum > mid:
                    count += 1
                    curSum = num
            return count <= k

        left, right = 0, len(nums) * max(nums)
        while left <= right:
            mid = (left + right) // 2
            if check(mid):
                right = mid - 1
            else:
                left = mid + 1
        return left


assert Solution().splitArray([7, 2, 5, 10, 8], 2) == 18
